pkg/cachemanager/cachemanager_test/cachemanager_integration_test.go

```diff
@@ -0,0 +1,333 @@
+	"github.com/open-policy-agent/gatekeeper/v3/test/testutils"
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+const (
+	eventuallyTimeout = 10 * time.Second
+	eventuallyTicker  = 2 * time.Second
+)
+
+var cfg *rest.Config
+
+func TestMain(m *testing.M) {
+	testutils.StartControlPlane(m, &cfg, 3)
+}
+
+// TestCacheManager_replay_retries tests that we can retry GVKs that error out in the reply goroutine.
+func TestCacheManager_replay_retries(t *testing.T) {
+	mgr, wm := testutils.SetupManager(t, cfg)
+	c := testclient.NewRetryClient(mgr.GetClient())
+
+	failPlease := make(chan string, 2)
+	reader := fakes.HookReader{
+		Reader: c,
+		ListFunc: func(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+			// Return an error the first go-around.
+			var failKind string
+			select {
+			case failKind = <-failPlease:
+			default:
+			}
+			if failKind != "" && list.GetObjectKind().GroupVersionKind().Kind == failKind {
+				return fmt.Errorf("synthetic failure")
+			}
+			return c.List(ctx, list, opts...)
+		},
+	}
+
+	cacheManager, dataStore, _, ctx := cacheManagerForTest(t, mgr, wm, reader)
+
+	cfClient, ok := dataStore.(*cachemanager.FakeCfClient)
+	require.True(t, ok)
+
+	// seed one gvk
+	configMapGVK := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "ConfigMap"}
+	cm := unstructuredFor(configMapGVK, "config-test-1")
+	require.NoError(t, c.Create(ctx, cm), "creating ConfigMap config-test-1")
+
+	podGVK := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Pod"}
+	pod := unstructuredFor(podGVK, "pod-1")
+	require.NoError(t, c.Create(ctx, pod), "creating Pod pod-1")
+
+	syncSourceOne := aggregator.Key{Source: "source_a", ID: "ID_a"}
+	require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceOne, []schema.GroupVersionKind{configMapGVK, podGVK}))
+
+	expected := map[cachemanager.CfDataKey]interface{}{
+		{Gvk: configMapGVK, Key: "default/config-test-1"}: nil,
+		{Gvk: podGVK, Key: "default/pod-1"}:               nil,
+	}
+
+	require.Eventually(t, expectedCheck(cfClient, expected), eventuallyTimeout, eventuallyTicker)
+
+	// set up a scenario where the list from replay will fail a few times
+	failPlease <- "ConfigMapList"
+	failPlease <- "ConfigMapList"
+
+	// this call should schedule a cache wipe and a replay for the configMapGVK
+	require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceOne, []schema.GroupVersionKind{configMapGVK}))
+
+	expected2 := map[cachemanager.CfDataKey]interface{}{
+		{Gvk: configMapGVK, Key: "default/config-test-1"}: nil,
+	}
+	require.Eventually(t, expectedCheck(cfClient, expected2), eventuallyTimeout, eventuallyTicker)
+
+	// cleanup
+	require.NoError(t, c.Delete(ctx, cm), "creating ConfigMap config-test-1")
```

Comment:
Will these be run if the test fails?

Reply:
good question! IIUC, these won't be run if the test fails previously on a previous require. This can lead us to leak some state between the test runs even if tear down the manager as part of the cleanup.

If this is a major issue, I can try to fix it in this PR. But, in my experience running this suite, I've not seen this issue. Can I fix it as a follow up item in https://github.com/open-policy-agent/gatekeeper/issues/2929 ?

Reply:
reworked these w `t.Cleanup` now.